torch/csrc/jit/runtime/graph_iterator.h

```diff
@@ -5,113 +5,196 @@ namespace jit {
+      current_ = owning_node_it;
+      ++current_;
+      if (current_ == owning_node_block->nodes().end()) {
+        move_up(*owning_node_it);
+      }
+    } else {
+      throw std::runtime_error("Node should not have had any child blocks.");
+    }
+  }
 
-        if (!then_block_empty && else_block_empty) {
-          // Set current_ to {then_block, then_block.begin()} and push nothing
-          // to the stack since the else block is empty.
-          current_.first = then_block;
-          current_.second = then_block->nodes().begin();
-        } else if (then_block_empty && !else_block_empty) {
-          // Set current_ to {else_block, else_block.begin()} and push nothing
-          // to the stack since the current block is already on the stack.
-          current_.first = else_block;
-          current_.second = else_block->nodes().begin();
-        } else if (!then_block_empty && !else_block_empty) {
-          // Set current_ to {then_block, then_block.begin()} and push the
-          // else_block to the stack so that it will be processed after.
-          block_stack_.emplace_back(else_block, else_block->nodes().begin());
-          current_.first = then_block;
-          current_.second = then_block->nodes().begin();
-        }
-      } else if (node->kind() == prim::Loop || node->kind() == prim::With) {
-        auto* body_block = node->blocks().at(0);
+  // Moves to the next adjacent node or up in to the parent if that is not
+  // possible.
+  void move_next() {
+    auto block = current_->owningBlock();
+    Node* previous = *current_;
+
+    // Increment to the next node in the current block.
+    ++current_;
 
-        bool body_block_empty =
-            body_block->nodes().begin() == body_block->nodes().end();
+    // Check if we're at the end of the block. If so we need
+    // to move upwards (if it makes sense to).
+    if (current_ == block->nodes().end()) {
+      move_up(previous);
+    }
+  }
 
-        if (!body_block_empty) {
-          // If body_block is not empty, push the current block onto the stack
-          // to resume processing it later and set current_ to {body_block,
-          // body_block.begin()}.
-          block_stack_.emplace_back(current_.first, current_.second);
+  // Moves to the next node in the graph into children if it can.
+  void move_into() {
+    if (*current_ == nullptr) {
+      return;
+    }
 
-          current_.first = body_block;
-          current_.second = body_block->nodes().begin();
-        }
+    auto node = *current_;
+
+    // Check if we're currently on a node that contains sub-nodes.
+    if (node->kind() == prim::If) {
+      // If nodes can have two child blocks. From an If node
+      // in `move_into` we only enter the first block which contains
+      // elements. After processing of those elements is done, we move
+      // to the next block (the else block if it has nodes) from the
+      // `move_up` call.
+      auto* then_block = node->blocks().at(0);
```

Comment:
this block i think recreates some of the logic that exists elsewhere. could we set current just equal to the beginning of the then block and if then_block is empty call move_next on it ?

Reply:
I don't think this would work - if then doesn't have any nodes it would return the equivalent of a `nullptr` so the `move_next` function couldn't get a reference to the parent to `move_up`?

Reply:
Sorry, why can't we set `current_ = node->blocks().at(0)->nodes()->begin(); move_next(current_)` ? i still think that would work

Reply:
I updated this in the PR. Let me know if the new approach works for you.